### PR TITLE
Add an optional MLFlow logger

### LINKED ISSUE #99: Add possibility to use the MLFlowLogger
Currently, only the Lighthning `TensorBoardLogger` is used to log and track the trainings.

It could be interresting to add the possiblility to activate the `MLFlowLogger` to take advantage of some MLFlow features, like the Model Registry of an MLFlow server.

While the MLFlow UI is, in my opinion, less user friendly that the Tensorboard one, I propose to be able to activate optionally the `MLFlowLogger` in addition to the default `TensorBoardLogger`.

### PR DESCRIPTION
**Pull request by A669015**
This PR resolves #99 with the following additions:
- add the possibility to log with the `MLFlowLogger`, in addition to thre `TensorboardLogger`, using the `--mlflow_log` option of the `bin/train.py` CLI,
- log the trained model in the MLFlow fashion at the final step of `train.py`, if the  `--mlflow_log` option is activated,
- add logging of metrics at the test stage,
- document the MLFlow logger output configuration

### CODE REVIEW

**Review by LBerth — APPROVED** [review]
Nice job!
Out of curiosity, do you have a screenshot of what the MLFlow interface looks like ?

File: py4cast/plots.py [hunk 1]
```diff
@@ -602,10 +642,24 @@ def on_step_end(self, obj: "AutoRegressiveLightning", label: str = "") -> None:
                 for t_i, loss_map in enumerate(mean_spatial_loss)
             ]
             tensorboard = obj.logger.experiment
+            mlflow_logger_idx = [
+                i for i, l in enumerate(obj.loggers) if isinstance(l, MLFlowLogger)
+            ]
+            mlflow_logger = (
+                obj.loggers[mlflow_logger_idx[0]] if mlflow_logger_idx else None
+            )
```
Comment (colon3ltocard) [inline py4cast/plots.py:650]:
```suggestion
        mlflow_logger = next([o for o in obj.loggers if isinstance(o, MLFlowLogger)], None)
```

File: py4cast/plots.py [hunk 2]
```diff
@@ -515,6 +542,10 @@ def on_step_end(self, obj: "AutoRegressiveLightning", label: str = "") -> None:
         Make the summary figure
         """
         tensorboard = obj.logger.experiment
+        mlflow_logger_idx = [
+            i for i, l in enumerate(obj.loggers) if isinstance(l, MLFlowLogger)
+        ]
+        mlflow_logger = obj.loggers[mlflow_logger_idx[0]] if mlflow_logger_idx else None
```
Comment (colon3ltocard) [inline py4cast/plots.py:548]:
```suggestion
        mlflow_logger = next([o for o in obj.loggers if isinstance(o, MLFlowLogger)], None)
```

File: py4cast/plots.py [hunk 3]
```diff
@@ -390,15 +391,30 @@ def plot_map(
             ]
             # TODO : don't create all figs at once to avoid matplotlib warning
             tensorboard = obj.logger.experiment
+            mlflow_logger_idx = [
+                i for i, l in enumerate(obj.loggers) if isinstance(l, MLFlowLogger)
+            ]
+            mlflow_logger = (
+                obj.loggers[mlflow_logger_idx[0]] if mlflow_logger_idx else None
+            )
```
Comment (colon3ltocard) [inline py4cast/plots.py:399]:
```suggestion
            mlflow_logger = next([o for o in obj.loggers if isinstance(o, MLFlowLogger)], None)
```
This code is duplicated many times, a property in the lightning module might do the trick ?

File: py4cast/lightning.py [hunk 4]
```diff
@@ -721,11 +722,28 @@ def on_validation_epoch_end(self):
             dict_metrics.update(self.psd_plot_metric.compute())
             dict_metrics.update(self.rmse_psd_plot_metric.compute())
             dict_metrics.update(self.acc_metric.compute())
+
+            mlflow_logger_idx = [
+                i for i, l in enumerate(self.loggers) if isinstance(l, MLFlowLogger)
+            ]
+            mlflow_logger = (
+                self.loggers[mlflow_logger_idx[0]] if mlflow_logger_idx else None
+            )
```
Comment (colon3ltocard) [inline py4cast/lightning.py:731]:
```suggestion
            mlflow_logger = next(
                iter([o for o in self.loggers if isinstance(o, MLFlowLogger)]), None
            )
```
Why not putting that into a property or method of the lightining module ? It is duplicated twice already plus used in the tplots (and plots have access to the LM)

Comment (A669015) [reply py4cast/lightning.py:731]: Sounds an excellent idea !

**Review by colon3ltocard — APPROVED** [review]

**Pull request merged**